src/scf/newton_tools.jl

```diff
@@ -0,0 +1,233 @@
+function KrylovKit.orthogonalize!(v::T, x::AbstractVector,
+                                        alg::OrthogonalizeAndProject) where {T}
+    v, x = orthogonalize!(v, x, alg.orth)
+    v = reshape(v, size(alg.ψ))
+    v = pack(alg.projector(v, alg.ψ))::T
+    v, x
+end
+function KrylovKit.gklrecurrence(operator, U::OrthonormalBasis, V::OrthonormalBasis, β,
+                                 alg::OrthogonalizeAndProject)
+    u = U[end]
+    v = operator(u, true)
+    v = axpy!(-β, V[end], v)
+    # for q in V # not necessary if we definitely reorthogonalize next step and previous step
+    #     v, = orthogonalize!(v, q, ModifiedGramSchmidt())
+    # end
+    α = norm(v)
+    rmul!(v, inv(α))
+
+    r = operator(v, false)
+    r = axpy!(-α, u, r)
+    for q in U
+        r, = orthogonalize!(r, q, alg)
+    end
+    β = norm(r)
+    return v, r, α, β
+end
+
+############################# OPERATORS ########################################
+
+"""
+    apply_Ω(basis::PlaneWaveBasis, δφ, φ, H, λ)
+
+Compute the application of Ω to an element on the tangent space.
+Here, an element on the tangent space can be written as
+      δP = Σ |φi><δφi| + hc
+where the δφi are of size Nb and are all orthogonal to the φj, 1 <= j <= N.
+Therefore we store them in the same kind of array than φ, with
+δφ[ik][:,i] = δφi for each k-point.
+Therefore, computing Ωδφ can be done analitically with the formula
+      Ωδφi = H*δφi - Σj <δφj|H|δφi> δφj - Σj λji δφj
+where λij = <φi|H|φj>.
+"""
+function apply_Ω(basis::PlaneWaveBasis, δφ, φ, H, λ)
+    Nk = length(basis.kpoints)
+    Ωδφ = similar(φ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        λk = λ[ik]
+        Ωδφk = similar(δφk)
+
+        N1 = size(δφk,2)
+        N2 = size(φk,2)
+        @assert N1 == N2
+        N = N1
+
+        # compute component on i
+        for i = 1:N
+            Hδφki = H.blocks[ik] * δφk[:,i]
+            Ωδφk[:,i] = Hδφki
+            for j = 1:N
+                Ωδφk[:,i] -= (φk[:,j]'Hδφki) * φk[:,j]
+                Ωδφk[:,i] -= λk[j,i] * δφk[:,j]
+            end
+        end
+        Ωδφ[ik] = Ωδφk
+    end
+    # ensure proper projection onto the tangent space
+    proj(Ωδφ, φ)
+end
+
+"""
+    apply_K(basis::PlaneWaveBasis, δφ, φ, ρ, occ)
+
+Compute the application of K to an element in the tangent space
+with the same notations as for the computation of Ω, we have
+      Kδφi = Π(dV * φi),
+Π being the projection on the space tangent to the φ and
+      dV = kernel(δρ)
+where δρ = Σi φi*conj(δφi) + hc.
+"""
+function apply_K(basis::PlaneWaveBasis, δφ, φ, ρ, occ)
+    Nk = length(basis.kpoints)
+
+    δρ = compute_density(basis, φ, δφ, occ)
+    dV = apply_kernel(basis, δρ; ρ=ρ)
+    Kδφ = similar(φ)
+
+    for ik = 1:Nk
+        kpt = basis.kpoints[ik]
+        φk = φ[ik]
+        dVφk = similar(φk)
+
+        for i = 1:size(φk,2)
+            φk_r = G_to_r(basis, kpt, φk[:,i])
+            dVφk_r = dV .* φk_r
+            dVφk_r = total_density(dVφk_r)
```

Comment:
Do this properly for spin (and test it!). V is a N x N x N x 2 array, and you should just multiply phi with V[:,:,:,kpt.spin]

Reply:
Okay ! how do I set the model to have spin ?

Reply:
See the tutorials

Reply:
Yes but the only one I found uses temperature https://juliamolsim.github.io/DFTK.jl/stable/examples/collinear_magnetism/ which is not supported yet. Can we talk about that in the afternoon ?